Gitpodify Couchers wrap up

darrenvong opened the pull request:
<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->
Taking over and wrapping up #1749. Added some instructions on using Gitpod in the frontend readme as another option for getting started!

Thanks @pawlean for introducing this to us in the first place :D

Open and test this at: https://gitpod.io/#https://github.com/Couchers-org/couchers/tree/backend/feature/gitpod-test

<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on frontend)
If you need help with any of these, please ask :)
--->
N/A on checklists, since no code changes to either frontend nor backend


<!---
Remember to request review from couchers-org/frontend, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

Commit:
Gitpodify Couchers

pawlean committed:
Revert "Fix .gitpod.yml"
This reverts commit beea2202b5a294ddb3982a1261d28fcfa47b47c1.

aapeliv committed:
Tweak .gitpod.yml

aapeliv committed:
Add instruction terminal that waits for frontend and backend

aapeliv committed:
Tweak gitpod

aapeliv committed:
Make ports visible

darrenvong committed:
Don't override API URL for use with companion app

darrenvong committed:
Tweak frontend start option to wait for backend port first

darrenvong committed:
Add documentation on how to use Gitpod in frontend README

darrenvong committed:
Don't open preview

As this is confusing and new joiners might think to use the preview.
Instead, they should be going to localhost:3000 in a new tab

aapeliv commented:
Is there really no way to do this without the local companion app? The whole point of gitpodify would be to be able to run it in their VM and have a completely standardized environment with no local setup needed.

(Also, that link is 404-ing.)

darrenvong commented:
Unless there is a way to workaround the fact cookies aren't being set properly when it comes from `8888-xxx.gitpod.io` to `3000-xxx.gitpod.io`... I don't think so. Maybe setting `COOKIE_DOMAIN` to `gitpod.io` in the backend environment variable? But that's part of the docker compose so I don't know how we can override that without changing that too, or make a Gitpod copy of docker-compose.

darrenvong commented:
> (Also, that link is 404-ing.)

The Gitpod badge link? It works for me (takes you to a login page, but should be fine once logged in with GitHub)

aapeliv commented:
> Maybe setting `COOKIE_DOMAIN` to `gitpod.io` in the backend environment variable?

That should be pretty straightforward, I think.

> > (Also, that link is 404-ing.)
> 
> The Gitpod badge link? It works for me (takes you to a login page, but should be fine once logged in with GitHub)

This link was failing: https://www.gitpod.io/blog/local-app. Either it was down very briefly (it's back up now), or they're doing some odd stuff with SSR and it defaults to a 404! I was getting that on all docs pages for a sec.

darrenvong commented:
> That should be pretty straightforward, I think.

Can it be done without creating another docker-compose and .env file though? Also I think the media stuff will still be broken, but that's broken locally anyway so maybe less of a big deal

aapeliv commented:
Yeah I think you can just add env vars before the docker-compose command.

darrenvong committed:
Try changing cookie domain to gitpod.io?

aapeliv commented:
No, actually; that probably won't work. Sorry, I misunderstood something.

I think we can create an override docker-compose file though. Something like `docker-compose.gitpod.yml` which only contains the overrides; then you can do something like `docker-compose -f docker-compose.yml -f docker-compose.gitpod.yml ...` and it should override the stuff specified in the latter file: https://docs.docker.com/compose/extends/#multiple-compose-files

darrenvong committed:
override COOKIE_DOMAIN with separate docker-compose yml

darrenvong committed:
Remove mentions of the companion app

darrenvong commented:
That extra `docker-compose.gitpod.yml` override did the trick and fixed the cookies issue! 🚀  So now it's all working within the browser:

https://gitpod.io/#https://github.com/Couchers-org/couchers/tree/backend/feature/gitpod-test

I've also updated the README and removed the reference to the companion app, so now it is really all on in the browser!

aapeliv commented on readme.md:
```diff
@@ -18,12 +20,12 @@ If you are a developer, please read [docs/contributing.md](docs/contributing.md)
+- Developer Contributor Guide: [docs/contributing.md](docs/contributing.md) explains how we work with this codebase and helps you get up and running with development.
+- CI/CD goodies: [docs/cicd.md](docs/cicd.md) lists the different CI/CD functionality that we've got set up and how you can use this to simplify life
+- Database & migrations: [docs/database.md](docs/database.md) explains how to use the database and how we version the database with Alembic and update it when new tables or columns are added
+- Docker: [docs/docker.md](docs/docker.md) has some tips on how to work with docker and be less miserable
```
Please don't replace change the `*` with `-`.

aapeliv commented on app/frontend/README.md:
```diff
@@ -75,4 +88,3 @@ Runs storybook, good for testing and developing components in isolation.
 ### `yarn build`
 
 Builds the app for production to the `build` folder.
-
```
Missing new line at EOF

aapeliv commented on app/docker-compose.yml:
```diff
@@ -41,4 +41,3 @@ services:
       - 5000:5000
     depends_on:
       - backend
-
```
Missing new line at EOF

aapeliv commented on .gitpod.yml:
```diff
@@ -0,0 +1,27 @@
+    before: cd app/frontend
+    init: |
+      yarn install
+      gp await-port 8888
```
Is this needed anymore? It should be able to `yarn start` before the backend is up?

darrenvong replied:
Yeah, but the build is usually a lot slower than the frontend dev server so if the preview is up and people click around at that in between point, it'll error until the backend finish building, so I thought it might be better to wait for it to be up first before starting the frontend

aapeliv replied:
Ahh, fair enough. It just adds a bit, I guess since it'll only actually build the frontend when you do `yarn start`?

aapeliv replied:
But yeah; I don't mind leaving it in!

aapeliv commented on .gitpod.yml:
```diff
@@ -0,0 +1,27 @@
+
+  - name: Instructions
+    before: cd app/
+    init: until (curl --head -s localhost:8888 localhost:8888 > /dev/null); do sleep 1; clear; echo "Waiting for backend & frontend to come up. This might take a few minutes!"; done
```
I think one of these should be `localhost:3000`?

aapeliv reviewed (dismissed):
Getting very nitpicky... sorry!

aapeliv commented on .gitpod.yml:
```diff
@@ -0,0 +1,27 @@
+  - name: Instructions
+    before: cd app/
+    init: until (curl --head -s localhost:8888 localhost:8888 > /dev/null); do sleep 1; clear; echo "Waiting for backend & frontend to come up. This might take a few minutes!"; done
+    command: echo "Success! You can now launch the frontend by going to http://localhost:3000/"
```
Also; I think this should be `gp url 3000`, not localhost?

darrenvong replied:
Ahh yes, I forgot to change it back when I was testing with the companion app, oops

aapeliv commented:
Just tried running this. It's looking good!

A few more things that I noticed:

* Map isn't loading because the dev key isn't set to allow `.gitpod.io`: I'll see if I can fix this one up
* In the same way you set `COOKIE_DOMAIN`, you should set `BASE_URL` to `gp url 3000`, so that e.g. signup links point to the right place. This might be a tad more tricky :/

aapeliv commented:
> Map isn't loading because the dev key isn't set to allow `.gitpod.io`: I'll see if I can fix this one up

Tried to fix this

darrenvong commented:
> > Map isn't loading because the dev key isn't set to allow `.gitpod.io`: I'll see if I can fix this one up
> 
> Tried to fix this

Is that a setting in some couchers mapbox account? Any luck?

> In the same way you set COOKIE_DOMAIN, you should set BASE_URL to gp url 3000, so that e.g. signup links point to the right place. This might be a tad more tricky :/

Yeah I think that might be difficult without some command to substitute in the value into the file itself before spinning the Docker containers up, but then people might end up accidentally committing those changes so I'm not sure that's a good idea :/ lol maybe the companion app is still the easiest way due to how finicky our setup is / Gitpod's domains are too dynamic

aapeliv commented:
> > In the same way you set COOKIE_DOMAIN, you should set BASE_URL to gp url 3000, so that e.g. signup links point to the right place. This might be a tad more tricky :/
> 
> Yeah I think that might be difficult without some command to substitute in the value into the file itself before spinning the Docker containers up, but then people might end up accidentally committing those changes so I'm not sure that's a good idea :/ lol maybe the companion app is still the easiest way due to how finicky our setup is / Gitpod's domains are too dynamic

I think if you add to the gitpod override docker compose file like this

```
version: "3"
services:
  backend:
    environment:
      - BASE_URL
```

with no value, then I think you should be able to pass `BASE_URL` from the command line, by doing `BASE_URL=$(gp url 3000) docker-compose ...`.

https://docs.docker.com/compose/environment-variables/#pass-environment-variables-to-containers

aapeliv committed:
Clean up

aapeliv committed:
Move gitpod button

aapeliv commented:
@darrenvong could you review please? (You made the pr so can't ask for review through gh)

aapeliv committed:
Try to add media base url

darrenvong commented on .gitpod.yml:
```diff
@@ -0,0 +1,27 @@
+      BASE_URL=$(gp url 3000) MEDIA_SERVER_BASE_URL=$(gp url 5000) docker-compose -f docker-compose.yml -f docker-compose.gitpod.yml up --build
+
+  - name: Install and launch the frontend with yarn
+    before: cd app/frontend
```
```suggestion
    before: cd app/web
```

And can we not rename `web` back to `frontend/blah` again... (unrelated rant)

darrenvong committed:
cd to `app/web` due to recent folder rename

darrenvong commented:
Just tried with my latest tweaks and this looks good to me! Can't approve it as I originally opened the PR 😅

lucaslcode reviewed (approved):
Works for me :)
